sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala

```diff
@@ -68,6 +69,15 @@ case class InMemoryRelation(
 
   override protected def innerChildren: Seq[SparkPlan] = Seq(child)
 
+  override def doCanonicalize(): logical.LogicalPlan =
+    copy(output = output.map(QueryPlan.normalizeExprId(_, child.output)),
+      storageLevel = StorageLevel.NONE,
+      child = child.canonicalized,
+      tableName = None)(
+      _cachedColumnBuffers,
+      sizeInBytesStats,
+      statsOfPlanToCache)
```

Comment:
We need to copy these cached data?

Reply:
`_cachedColumnBuffers` can't be null. If it is null, `copy` will trigger `buildBuffers`.

Reply:
`statsOfPlanToCache` and `sizeInBytesStats`, too? For instance, `ResolveHint`  drops hints in canonicalization: https://github.com/apache/spark/blob/3675af7247e841e9a689666dc20891ba55c612b3/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/hints.scala#L44

Reply:
`cachedColumnBuffers`, `sizeInBytesStats`, `statsOfPlanToCache` won't be considered when comparing two `InMemoryRelation`. So instead of create empty instances of statistics, I just use the original values.

Reply:
aha, ok. Thanks!